### PR TITLE
GH-39265: [Java] Make it run well with the netty newest version 4.1.104

### LINKED ISSUE #39265: [Java] Make it run well with the netty newest version 4.1.104
### Describe the enhancement requested

When I used `netty arrow memory 14.0.1` and `netty 4.1.104.Final` in Spark, the following error occurred,
After pr: https://github.com/netty/netty/pull/13613, `PoolArena` no longer extends `SizeClasses`, but instead uses it as one of its fields, as follows:
image
in order to ensure that `netty arrow memory 14.0.1` works well with `netty 4.1.104.Final` version, I suggest making similar modifications here.
1.Compilation errors are as follows:
https://github.com/panbingkun/spark/actions/runs/7237466030/job/19717162391
image

2.Some bugs have been fixed in `netty 4.1.104.Final` as follows:
image
image

4.1.104.Final release note: https://netty.io/news/2023/12/15/4-1-104-Final.html
4.1.103.Final release note: https://netty.io/news/2023/12/13/4-1-103-Final.html
4.1.101.Final release note: https://netty.io/news/2023/11/09/4-1-101-Final.html

### Component(s)

Java

### PR DESCRIPTION
**Pull request by panbingkun**
### Describe the enhancement requested

When I used `netty arrow memory 14.0.1` and `netty 4.1.104.Final` in Spark, the following error occurred,
After pr: https://github.com/netty/netty/pull/13613, `PoolArena` no longer extends `SizeClasses`, but instead uses it as one of its fields, as follows:
image
in order to ensure that `netty arrow memory 14.0.1` works well with `netty 4.1.104.Final` version, I suggest making similar modifications here.
1.Compilation errors are as follows:
https://github.com/panbingkun/spark/actions/runs/7237466030/job/19717162391
image

2.Some bugs have been fixed in `netty 4.1.104.Final` as follows:
image
image

4.1.104.Final release note: https://netty.io/news/2023/12/15/4-1-104-Final.html
4.1.103.Final release note: https://netty.io/news/2023/12/13/4-1-103-Final.html
4.1.101.Final release note: https://netty.io/news/2023/11/09/4-1-101-Final.html

### Component(s)

Java
* Closes: #39265

### CODE REVIEW

**Comment by lidavidm** [comment]
Hmm, dropping compatibility with older versions of Netty isn't great either, if people are using older gRPC. But we might not have a choice. CC @jduo

**Comment by lidavidm** [comment]
CC @danepitkin this means we will have a hard compatibility break between netty versions

**Comment by danepitkin** [comment]
I wonder if it is possible to support both old and new netty versions, maybe using run time reflection?

**Comment by panbingkun** [comment]
I think using the reflection mechanism can solve compatibility issues with older versions, but there may be some `performance losses`. Can we accept this?

**Review by jbonofre — APPROVED** [review]
The change looks good to me. I would like to test in Flight context to be sure it's clear with gRPC.

**Comment by lidavidm** [comment]
Well, we already use reflection here so that's moot.

**Comment by lidavidm** [comment]
That said, it seems there's a different way to get chunk size? If we can validate that they're the same, why not use that instead and avoid poking internals? https://github.com/netty/netty/blob/427ace4cdd86cce84e467ec0a8a81ca022ed6f72/buffer/src/main/java/io/netty/buffer/PooledByteBufAllocator.java#L663-L671

**Comment by jbonofre** [comment]
@lidavidm yeah, that's my point about buffer allocator. I know gRPC is using that internally as well. It makes sense to use the "regular" netty API instead of tweaking by the internals. I can work on this if you want 😄 (I have some PRs on the way 😄 ).

**Comment by lidavidm** [comment]
@jbonofre if you have the interest then by all means. It is much appreciated.

**Comment by jbonofre** [comment]
@lidavidm my pleasure ! Thanks !

**Comment by panbingkun** [comment]
From a coding perspective, `PooledByteBufAllocator#metric()#chunkSize()` and `PooledByteBufAllocator#directArenas(0)#chunkSize()` are indeed the same, as follows:

1.PooledByteBufAllocator#metric()#chunkSize()
https://github.com/netty/netty/blob/427ace4cdd86cce84e467ec0a8a81ca022ed6f72/buffer/src/main/java/io/netty/buffer/PooledByteBufAllocator.java#L569-L571
image

https://github.com/netty/netty/blob/427ace4cdd86cce84e467ec0a8a81ca022ed6f72/buffer/src/main/java/io/netty/buffer/PooledByteBufAllocator.java#L330
image

https://github.com/netty/netty/blob/427ace4cdd86cce84e467ec0a8a81ca022ed6f72/buffer/src/main/java/io/netty/buffer/PooledByteBufAllocatorMetric.java#L30-L32
image

https://github.com/netty/netty/blob/427ace4cdd86cce84e467ec0a8a81ca022ed6f72/buffer/src/main/java/io/netty/buffer/PooledByteBufAllocatorMetric.java#L96-L98
image

https://github.com/netty/netty/blob/427ace4cdd86cce84e467ec0a8a81ca022ed6f72/buffer/src/main/java/io/netty/buffer/PooledByteBufAllocator.java#L283
image

https://github.com/netty/netty/blob/427ace4cdd86cce84e467ec0a8a81ca022ed6f72/buffer/src/main/java/io/netty/buffer/PooledByteBufAllocator.java#L669-L671
image

File: java/memory/memory-netty/src/main/java/io/netty/buffer/PooledByteBufAllocatorL.java [hunk 1]
```diff
@@ -137,7 +137,6 @@ private class InnerAllocator extends PooledByteBufAllocator {
 
     private final PoolArena<ByteBuffer>[] directArenas;
     private final MemoryStatusThread statusThread;
-    private final int chunkSize;
```
Comment (panbingkun) [inline java/memory/memory-netty/src/main/java/io/netty/buffer/PooledByteBufAllocatorL.java:140]: From code analysis, it seems that the `chunkSize` field here is useless. After the constructor is assigned, it is no longer used, and this value is directly derived from the parent class.

File: java/memory/memory-netty/src/main/java/io/netty/buffer/PooledByteBufAllocatorL.java [hunk 2]
```diff
@@ -166,7 +163,7 @@ private UnsafeDirectLittleEndian newDirectBufferL(int initialCapacity, int maxCa
 
       if (directArena != null) {
 
-        if (initialCapacity > directArena.chunkSize) {
+        if (initialCapacity > chunkSize()) {
```
Comment (panbingkun) [inline java/memory/memory-netty/src/main/java/io/netty/buffer/PooledByteBufAllocatorL.java:166]: The value of `directArena.chunkSize` is actually equal to `chunkSize()`, and it is a final.

**Comment by panbingkun** [comment]
After completing the above changes, I found that it can run well with both `netty version 4.1.104` and `netty version 4.1.100`.

File: java/memory/memory-netty/src/main/java/io/netty/buffer/PooledByteBufAllocatorL.java (hunk 2 above)
Comment (lidavidm) [reply java/memory/memory-netty/src/main/java/io/netty/buffer/PooledByteBufAllocatorL.java:166]: Thanks.  One final thing, I think chunkSize is deprecated in Netty? Is it possible to use the non-deprecated version?

**Comment by danepitkin** [comment]
Thanks for finding a way to support old versions of netty!

**Pull request merged**